Example/Tests/Caching/MHVThingCacheQueryTests.m

```diff
@@ -0,0 +1,258 @@
+//
+//  MHVThingCacheQueryTests.m
+//  healthvault-ios-sdk
+//
+//  Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import <XCTest/XCTest.h>
+#import "MHVMockDatabase.h"
+#import "MHVThingCache.h"
+#import "Kiwi.h"
+
+static NSString *kRecordUUID = @"11111111-aaaa-aaaa-aaaa-111111111111";
+
+SPEC_BEGIN(MHVThingCacheQueryTests)
+
+describe(@"MHVThingCache", ^
+{
+    __block MHVMockDatabase *database;
+    __block MHVThingCache *thingCache;
+    __block MHVThingQueryResultCollection *returnedCollection;
+    __block NSError *returnedError;
+    
+    MHVPersonInfo *testPerson = [MHVPersonInfo new];
+    testPerson.records = [MHVRecordCollection new];
+    
+    MHVRecord *record = [MHVRecord new];
+    record.ID = [[NSUUID alloc] initWithUUIDString:kRecordUUID];
+    [testPerson.records addObject:record];
+    
+    MHVThingCacheConfiguration *cacheConfig = [MHVThingCacheConfiguration new];
+    cacheConfig.cacheTypeIds = @[[MHVAllergy typeID],
+                                 [MHVWeight typeID]];
+    
+    KWMock<MHVConnectionProtocol> *mockConnection = [KWMock mockForProtocol:@protocol(MHVConnectionProtocol)];
+    [mockConnection stub:@selector(cacheConfiguration) andReturn:cacheConfig];
+    [mockConnection stub:@selector(personInfo) andReturn:testPerson];
+    [mockConnection stub:@selector(thingClient) andReturn:nil];
+    [mockConnection stub:@selector(applicationId) andReturn:@"app-id-1"];
+    
+    beforeEach(^
+               {
+                   database = nil;
+                   thingCache = nil;
+                   returnedCollection = nil;
+                   returnedError = nil;
+               });
+    
+    context(@"when query record has not synced", ^
```

Comment:
I think we can still do a better job naming tests. The point of BDD is that the name of the test provides a clear explanation of exactly what is failing, where it's failing and why. The problem with the name of many of these tests is that a developer will need to hunt around the project to figure out what was tested and eventually why it's broken.

For example, if this particular test fails this is what we see: **MHVThingCache_WhenQueryRecordHasNotSynced_ThingCollectionShouldBeNil**. I'm not exactly sure what method I called that caused this error, which thing collection we are talking about, and what a query record is

Instead maybe the name should be "when cachedResultsForQueries is called and the cache is not ready" and it "should provide a nil result collection" and it "should provide a detailed error" (We should be caching that the error is the cone we expect).

**MHVThingCache_WhenCachedResultsForQueriesIsCalledAndTheCacheIsNotReady_ShouldProvideANilResultCollection**